app/src/main/java/app/revanced/integrations/shared/checks/CheckEnvironmentPatch.java

```diff
@@ -0,0 +1,294 @@
+     * <br>
+     * If the app was installed at a later time, it is likely, the app was downloaded pre-patched, the user
+     * waited too long to install the app or the patch time is too long ago.
+     */
+    private static final Check isNearPatchTimeCheck = new Check(
+            "revanced_check_environment_not_near_patch_time"
+    ) {
+        @Override
+        protected boolean run() {
+            final var duration = System.currentTimeMillis() - PATCH_TIME;
+
+            final var passed = duration < 1000 * 60 * 30; // 1000 ms * 60 s * 30 min
+
+            if (passed) {
+                Logger.printDebug(() -> "Installed in " + duration / 1000 + " seconds after patch");
+            }
+
+            return passed;
+        }
+    };
+
+    /**
+     * Check if the IP address is the same as during the patch.
+     * <br>
+     * If the IP address is the same as during the patch, it is likely, the app was just patched by the user.
+     * <br>
+     * If the IP address is different, the app was likely downloaded pre-patched or the patch time is too long ago.
+     */
+    private static final Check hasPatchTimePublicIPCheck = new Check(
+            "revanced_check_environment_not_patch_time_public_ip"
+    ) {
+        @Override
+        protected boolean run() {
+            // Using multiple services to increase reliability, distribute the load and minimize tracking.
+            final var getIpServices = new ArrayList<String>() {
+                {
+                    add("https://wtfismyip.com/text");
+                    add("https://whatsmyfuckingip.com/text");
+                    add("https://api.ipify.org?format=text");
+                    add("https://icanhazip.com");
+                    add("https://ifconfig.me/ip");
+                }
+            };
+
+            String publicIP = null;
+
+            try {
+                var service = getIpServices.get(new Random().nextInt(getIpServices.size() - 1));
+
+                Logger.printDebug(() -> "Using " + service + " to get public IP");
+
+                HttpURLConnection urlConnection = (HttpURLConnection) new URL(service).openConnection();
+
+                try {
+                    BufferedReader in = new BufferedReader(new InputStreamReader(urlConnection.getInputStream()));
+
+                    publicIP = in.readLine();
+
+                    String finalPublicIP = publicIP;
+                    Logger.printDebug(() -> "Got public IP " + finalPublicIP);
+
+                    in.close();
+                } finally {
+                    urlConnection.disconnect();
+                }
+            } catch (Exception e) {
+                // If the app does not have the INTERNET permission or the service is down,
+                // the public IP can not be retrieved.
+                Logger.printDebug(() -> "Failed to get public IP address: " + e.getMessage());
```

Comment:
For some reason seing this. e.getMessage() is null.

Reply:
Thats right, but this type of message is a debug log. Could add the same overload for that as well?